<h1>Activity 2: UK_ACCIDENT_DATA OVERVIEW</h1>
<p>Analyst: Arnel N. Caparros</p>
<hr>

In [67]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')
# import matplotlib.pyplot as plt
# import seaborn as sns

In [68]:
accidents = pd.read_csv('uk_accident_data.csv')

In [69]:
accidents

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    05/06/2019  51.506187   
1       200701BS65737           Serious    02/07/2019  51.495029   
2       200701BS66127           Serious    26/08/2019  51.517715   
3       200701BS66128           Serious    16/08/2019  51.495478   
4       200701BS66837            Slight    03/09/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18/02/2022  57.374005   
660675  201091NM01881            Slight    21/02/2022  57.232273   
660676  201091NM01935            Slight    23/02/2022  57.585044   
660677  201091NM01964           Serious    23/02/2022  57.214898   
660678  201091NM02142           Serious    28/02/2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

<h2>Exploratory Data Analytics (EDA)</h2>
<hr>

<h1>1. How many total accidents are recorded in the dataset?</h1>

In [70]:
total_accidents = accidents.shape[0]

In [71]:
total_accidents

660679

<p>Insight NO.1: The total accidents are recorded in the dataset is "660679"</p>
<hr>

<h1>2.What Light_Conditions do most accidents occur?</h1>

In [72]:
accidents['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

<p>Insight NO.2: The Light_Conditions with the most accidents occur was Daylight</p>
<hr>

<h1>3. What is the distribution of accidents by severity?</h1>

In [73]:
accidents['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

<p>Insight NO.3: The distribution of accidents by severity are:</p>
Slight     </p>
Serious     </p>
Fatal        </p>
<hr>

<h1>4. Which severity category has the highest number of accidents?</h1>

In [74]:
accidents['Accident_Severity'].max()

'Slight'

<p>Insight NO.4: The severity category with the highest number of accidents is "Slight"</p>
<hr>

<h1>5. Which District Area has the most accidents?</h1> 

In [75]:
accidents['District Area'].max()

'York'

<p>Insight NO.5: The District Area with the most accidents is "York"</p>
<hr>

<h1>6. Are urban areas riskier than rural areas in terms of accident counts?</h1>

In [76]:
accidents['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          421663
Rural          238990
Unallocated        11
Name: count, dtype: int64

<p>Insight NO.6: The data shows that it is riskier to Urban to rural</p>
<hr>

<h1>7. What Road_Surface_Conditions has the most accidents?</h1>

In [77]:
accidents['Road_Surface_Conditions'].value_counts()

Road_Surface_Conditions
Dry                     447821
Wet or damp             186708
Frost or ice             18517
Snow                      5890
Flood over 3cm. deep      1017
Name: count, dtype: int64

<p>Insight NO.7: The data shows that the Road_Surface_Conditions with the highest accidents is "Dry"</p>
<hr>

<h1>8. How many Number_of_Casualties are recorded in the dataset?
</h1>

In [78]:
accidents['Number_of_Casualties'].sum()

np.int64(896568)

<p>Insight NO.7: The Number_of_Casualties that recorded in the dataset is "896568"</p>
<hr>

<h1>9. What is the max Number_of_Casualties that recorded in the dataset? ¶</h1>

In [79]:
accidents['Number_of_Casualties'].max()

np.int64(68)

<p>Insight NO.7: The max Number_of_Casualties that recorded in the dataset is "68"</p>
<hr>

<h1>10. Do accidents occur more frequently on single carriageways, dual carriageways, or motorways?</h1>

In [80]:
accidents['Road_Type'].value_counts()

Road_Type
Single carriageway    492143
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
Name: count, dtype: int64

<p>Insights NO. 10: The data shows that in the "Single carriageway", accidents occurs more</p>
<hr>

<h1>11. Is there a correlation between Number_of_Casualties and Number_of_Vehicles?</h1>

In [81]:
accidents[['Number_of_Casualties','Number_of_Vehicles']].corr()

Number_of_Casualties  Number_of_Vehicles
Number_of_Casualties              1.000000            0.228889
Number_of_Vehicles                0.228889            1.000000

<p>Insights NO. 11: It shows that as the number of vehicles increases, casualties tend to increase slightly, but the relationship is weak.</p>
<hr>

<h1>12. How many vehicles are involved per Accident_Severity (Slight, Serious and Fatal)?</h1>

In [82]:
avg_by_severity = accidents.groupby("Accident_Severity")["Number_of_Vehicles"].sum()

In [83]:
avg_by_severity

Accident_Severity
Fatal        15477
Serious     148057
Slight     1046338
Name: Number_of_Vehicles, dtype: int64

<p>Insights NO. 12: The data shows the number of vehicles involved in Slight is "1046338", in Serious is "148057" and in Fatal is "15477"
</p><hr>

<h1>13. What is the average Number_of_Casualties per accident by Urban_or_Rural_Area?</h1>

In [84]:
accidents.groupby("Urban_or_Rural_Area")["Number_of_Casualties"].mean()

Urban_or_Rural_Area
Rural          1.479204
Unallocated    1.181818
Urban          1.287815
Name: Number_of_Casualties, dtype: float64

<p>Insights NO. 13: The data shows that the average Number_of_Casualties in Urban is "1.287815", in Rural "1.479204", and in Unallocated "1.181818"
</p>
<hr>

<h1>14. What Vehicle_Type is the most prone to accidents base on the data</h1>

In [85]:
accidents['Vehicle_Type'].value_counts()

Vehicle_Type
Car                                      497992
Van / Goods 3.5 tonnes mgw or under       34160
Bus or coach (17 or more pass seats)      25878
Motorcycle over 500cc                     25657
Goods 7.5 tonnes mgw and over             17307
Motorcycle 125cc and under                15269
Taxi/Private hire car                     13294
Motorcycle over 125cc and up to 500cc      7656
Motorcycle 50cc and under                  7603
Goods over 3.5t. and under 7.5t            6096
Other vehicle                              5637
Minibus (8 - 16 passenger seats)           1976
Agricultural vehicle                       1947
Pedal cycle                                 197
Data missing or out of range                  6
Ridden horse                                  4
Name: count, dtype: int64

<p>Insights NO. 14: The data shows that the Vehicle_Type with the most prone to accidents is "Car"</p>

<h1>15. What is the highest Number_of_Vehicles involved in an accidents?</h1>

In [86]:
accidents['Number_of_Vehicles'].max()

np.int64(32)

<p>Insights NO. 14: The highest Number_of_Vehicles involved in an accidents s "32"</p>
<hr>

<h1>16. Is there a correlation between latitude and longitude</h1>

In [87]:
accidents.dropna(subset=["Latitude","Longitude"])["Latitude"].corr(accidents["Longitude"])


np.float64(-0.3981136594753895)

<p>Insights NO. 16: As the data shows the correlation between lattitude and longitude is -0.3981136594753895, meaning there was no correlation between them.</p>
<hr>

<h1>17. How does the number of vehicles correlate with light conditions
</h1>

In [88]:
vehicles_light = accidents.groupby("Light_Conditions")["Number_of_Vehicles"].mean()
print(vehicles_light)

Light_Conditions
Darkness - lighting unknown    1.758482
Darkness - lights lit          1.775498
Darkness - lights unlit        1.752654
Darkness - no lighting         1.599781
Daylight                       1.865385
Name: Number_of_Vehicles, dtype: float64


<p>Insights NO. 17:</p>
Daylight has the highest average (1.87)</p>
Darkness - no lighting has the lowest (1.60)</p>
Other lighting conditions are similar (~1.75)</p>
It means that daylight is the most prone in accidents maybe because of traffic
<hr>

<h1>18. Which Road_Type has the highest number of accidents in urban vs rural areas?</h1>

In [89]:
accidents.groupby(["Urban_or_Rural_Area", "Road_Type"]).size().sort_values(ascending=False).head(1)


Urban_or_Rural_Area  Road_Type         
Urban                Single carriageway    324117
dtype: int64

<p>Insights NO. 17: Tha data shows that the Road_Type that has the highest number of accidents in urban vs rural areas is "Single carriageway"</p>
<hr>

<h1>19. What is the highest Number_of_Vehicles involved in an accidents?</h1>

In [90]:
accidents["District Area"].value_counts().head(1)

District Area
Birmingham    13491
Name: count, dtype: int64

<p>Insights NO. 19: The highest Number_of_Vehicles involved in an accidents is "Birmingham" with "13491" accidents</p>
<hr>

<h1>20. Is there a same Latitude in the table?</h1>

In [91]:
duplicate_lat = accidents["Latitude"].duplicated().sum()

In [92]:
duplicate_lat

np.int64(149060)

<p>Insights NO. 20: Yes! There is a total of "149060" that has the same Latitude in the table</p>
<hr>

<h1>Extracting Date Information Using Pandas Date Time</h1>
<hr>

In [93]:
accidents['Accident Date'] = accidents['Accident Date'] = pd.to_datetime(accidents['Accident Date'], dayfirst = True, errors ='coerce')

In [94]:
accidents['Year']= accidents['Accident Date'].dt.year
accidents['Month']= accidents['Accident Date'].dt.month
accidents['Day']= accidents['Accident Date'].dt.day
accidents['DayOfWeek']= accidents['Accident Date'].dt.dayofweek #Monday = 0, #Sunday = 6

In [95]:
accidents.dtypes

Index                              object
Accident_Severity                  object
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                   object
District Area                      object
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions            object
Road_Type                          object
Urban_or_Rural_Area                object
Weather_Conditions                 object
Vehicle_Type                       object
Year                                int32
Month                               int32
Day                                 int32
DayOfWeek                           int32
dtype: object

<h1>Changing to Null Value</h1>

In [96]:
accidents['Index'] = accidents['Index'].fillna(accidents['Index'].mode()[0])
accidents['Accident_Severity'] = accidents['Accident_Severity'].fillna(accidents['Accident_Severity'].mode()[0])
accidents['Accident Date'] = accidents['Accident Date'].fillna(accidents['Accident Date'].mode()[0])
accidents['Latitude'] = accidents['Latitude'].fillna(accidents['Latitude'].mode()[0])
accidents['Light_Conditions'] = accidents['Light_Conditions'].fillna(accidents['Light_Conditions'].mode()[0])
accidents['District Area'] = accidents['District Area'].fillna(accidents['District Area'].mode()[0])
accidents['Longitude'] = accidents['Longitude'].fillna(accidents['Longitude'].mode()[0])
accidents['Number_of_Casualties'] = accidents['Number_of_Casualties'].fillna(accidents['Number_of_Casualties'].mode()[0])
accidents['Number_of_Vehicles'] = accidents['Number_of_Vehicles'].fillna(accidents['Number_of_Vehicles'].mode()[0])
accidents['Road_Surface_Conditions'] = accidents['Road_Surface_Conditions'].fillna(accidents['Road_Surface_Conditions'].mode()[0])
accidents['Road_Type'] = accidents['Road_Type'].fillna(accidents['Road_Type'].mode()[0])
accidents['Urban_or_Rural_Area'] = accidents['Urban_or_Rural_Area'].fillna(accidents['Urban_or_Rural_Area'].mode()[0])
accidents['Weather_Conditions'] = accidents['Weather_Conditions'].fillna(accidents['Weather_Conditions'].mode()[0])
accidents['Vehicle_Type'] = accidents['Vehicle_Type'].fillna(accidents['Vehicle_Type'].mode()[0])
accidents['Year'] = accidents['Year'].fillna(accidents['Year'].mode()[0])
accidents['Month'] = accidents['Month'].fillna(accidents['Month'].mode()[0])
accidents['Day'] = accidents['Day'].fillna(accidents['Day'].mode()[0])
accidents['DayOfWeek'] = accidents['DayOfWeek'].fillna(accidents['DayOfWeek'].mode()[0])

In [97]:
accidents.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
Year                       0
Month                      0
Day                        0
DayOfWeek                  0
dtype: int64

<h1>21-23.Which year recorded the highest number of accidents?</h1>
<h1>Which year recorded the lowest number of accidents?</h1>
<h1>How many accidents happened in each year?</h1>

In [121]:
accidents_per_year = accidents.groupby(accidents["Accident Date"].dt.year).size()
accidents_per_year

Accident Date
2019    182115
2020    170591
2021    163554
2022    144419
dtype: int64

<p>Insights NO. 21: The highest number af accidents by year is "2019"</p>
Insights NO. 22: The lowest number af accidents by year is "2022"</p>
Insights NO. 23: The number af accidents happened by year are:</p>
in 2019    182115,</p>
in 2020    170591,</p>
in 2021    163554, </p>
and 2022    144419</p>
<hr>

<h1>24. Are there more accidents on weekends than weekdays?</h1>

In [113]:
weekend = accidents[accidents["Accident Date"].dt.day_name().isin(["Saturday","Sunday"])].shape[0]
weekday = accidents[~accidents["Accident Date"].dt.day_name().isin(["Saturday","Sunday"])].shape[0]
print("Weekends" if weekend > weekday else "Weekdays")


Weekdays


<p>Insights NO. 24: It shows that there more accidents in weekdays than in weekends</p>
<hr>

<h1>25. Does the number of vehicles involved in accidents change over the years?</h1>

In [128]:
accidents.groupby(accidents["Accident Date"].dt.year)["Number_of_Vehicles"].mean()

Accident Date
2019    1.839310
2020    1.826615
2021    1.826229
2022    1.832273
Name: Number_of_Vehicles, dtype: float64

<p>Insights NO. 25: Yes, but not much, because the numbers slightly fluctuate but remain almost constant.</p>
<hr>

<p>

Which year recorded the lowest number of accidents?

What is the monthly trend of accidents across all years?

Do accidents peak in a particular month each year?

On which day of the week do most accidents occur?

Are there more accidents on weekends than weekdays?

What is the average number of casualties per month?

How do accidents vary by quarter of the year (Q1, Q2, Q3, Q4)?

Which year-month combination had the highest number of accidents?

Does the number of vehicles involved in accidents change over the years?</p>